Relax some limitations of InferenceMode.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55653 Move set_version_counter out of DifferentiableViewMeta constructor.
* #55572 Make VariableVersion::DISABLED the default constructor for VariableVersion.
* **#54403 Relax some limitations of InferenceMode.**

A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

Commit:
Allow mixing normal and inference tensor in normal mode as long as it
doesn't bump version.

[ghstack-poisoned]

Commit:
Update on "Allow mixing normal and inference tensor in normal mode as long as it"


doesn't bump version.

[ghstack-poisoned]

Commit:
Update on "Allow mixing normal and inference tensor in normal mode as long as it"


doesn't bump version.

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -88,6 +88,7 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
   if (c10::InferenceMode::is_enabled()) {
     // See Note [Expected TLS state in InferenceMode] for why we don't add Autograd & InplaceOrView keys.
     key_set_ = key_set;
+    version_counter_ = VariableVersion(-1);
```
Don't do it this way. If you represent missing VC as -1, you still have to do the dynamic allocation for the variable version intrusive ptr target. Instead, represent not present version counter as the intrusive ptr nullptr equivalent.

ezyang commented:
I think there's a correctness problem with erroring on bumping, versus erroring on read. From correspondence:

> It seems to me that mutating an inference tensor outside of inference mode is safe, as long as you haven't saved it for backwards.
>
> But the vice versa isn't safe: you can't save an inference tensor for backwards, because you might reenter an inference region (where inplace mutation could occur without bumping the VC at all)

Also, zeroing out the VC in TensorImpl constructor seems a bit smelly to me, because that's not where the final setting of VC gets done anyway for views (it can't be, because we need to share th VC in that situation)

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. Inference tensor can only be created inside InferenceMode. But not
all tensors created in InferenceMode are inference tensors. Namely view
of normal tensor inside InferenceMode produce a normal tensor, which is
exactly the same as creating a view inside NoGradMode.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be inplace updated outside InferenceMode.
4. It's not allowed to take views of inference tensor outside
InferenceMode. (This can be relaxed if needed. See comment in the
generated InplaceOrView_x.cpp.)

```
// Theorectically we can allow view ops on inference tensor in normal mode
// as long as we mark the output also inference tensor in this kernel.
// But it'll break an invariant we currently have: inference tensor can
// only be created inside InferenceMode.
// This invariant makes inference tensor easier for users to understand,
// so we should only break it when there's a valid use case.
TORCH_CHECK(!self.unsafeGetTensorImpl()->is_inference_tensor(),
  "Calling view ops on inference tensor outside InferenceMode is not allowed, ",
  "consider doing it inside infernce mode to work around this error. ",
  "If you have a valid use case, please make a feature request to PyTorch.");
```

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. Inference tensor can only be created inside InferenceMode. But not
all tensors created in InferenceMode are inference tensors. Namely view
of normal tensor inside InferenceMode produce a normal tensor, which is
exactly the same as creating a view inside NoGradMode.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be inplace updated outside InferenceMode.
4. It's not allowed to take views of inference tensor outside
InferenceMode. (This can be relaxed if needed. See comment in the
generated InplaceOrView_x.cpp.)

```
// Theorectically we can allow view ops on inference tensor in normal mode
// as long as we mark the output also inference tensor in this kernel.
// But it'll break an invariant we currently have: inference tensor can
// only be created inside InferenceMode.
// This invariant makes inference tensor easier for users to understand,
// so we should only break it when there's a valid use case.
TORCH_CHECK(!self.unsafeGetTensorImpl()->is_inference_tensor(),
  "Calling view ops on inference tensor outside InferenceMode is not allowed, ",
  "consider doing it inside infernce mode to work around this error. ",
  "If you have a valid use case, please make a feature request to PyTorch.");
```

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. Inference tensor can only be created inside InferenceMode. But not
all tensors created in InferenceMode are inference tensors. Namely view
of normal tensor inside InferenceMode produce a normal tensor, which is
exactly the same as creating a view inside NoGradMode.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be inplace updated outside InferenceMode.
4. It's not allowed to take views of inference tensor outside
InferenceMode. (This can be relaxed if needed. See comment in the
as_view() function in VariableTypeUtils.h.)

```
// Theorectically we can allow view ops on inference tensor in normal mode
// as long as we mark the output also inference tensor in this kernel.
// But it'll break an invariant we currently have: inference tensor can
// only be created inside InferenceMode.
// This invariant makes inference tensor easier for users to understand,
// so we should only break it when there's a valid use case.
TORCH_CHECK(!self.unsafeGetTensorImpl()->is_inference_tensor(),
  "Calling view ops on inference tensor outside InferenceMode is not allowed, ",
  "consider doing it inside infernce mode to work around this error. ",
  "If you have a valid use case, please make a feature request to PyTorch.");
```

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -222,29 +223,40 @@ struct C10_API NamedTensorMetaInterface {
+    // version is only used when enabled=true.
+    if (enabled) {
+      version_counter_ = c10::make_intrusive<VersionCounter>(version);
+    }
```
Just make two constructors LOL

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. Inference tensor can only be created inside InferenceMode. But not
all tensors created in InferenceMode are inference tensors. Namely view
of normal tensor inside InferenceMode produce a normal tensor, which is
exactly the same as creating a view inside NoGradMode.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be inplace updated outside InferenceMode.
4. It's not allowed to take views of inference tensor outside
InferenceMode. (This can be relaxed if needed. See comment in the
as_view() function in VariableTypeUtils.h.)

```
// Theorectically we can allow view ops on inference tensor in normal mode
// as long as we mark the output also inference tensor in this kernel.
// But it'll break an invariant we currently have: inference tensor can
// only be created inside InferenceMode.
// This invariant makes inference tensor easier for users to understand,
// so we should only break it when there's a valid use case.
TORCH_CHECK(!self.unsafeGetTensorImpl()->is_inference_tensor(),
  "Calling view ops on inference tensor outside InferenceMode is not allowed, ",
  "consider doing it inside infernce mode to work around this error. ",
  "If you have a valid use case, please make a feature request to PyTorch.");
```

[ghstack-poisoned]

Commit:
Update on "[WIP]Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. Inference tensor can only be created inside InferenceMode. But not
all tensors created in InferenceMode are inference tensors.1. All tensors created in InferenceMode are inference tensors except
   for view ops.
   - view ops produce output has the same is_inference_tensor property
     as their input. Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as
output.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.

[ghstack-poisoned]

Commit:
Update on "[WIP]Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view ops.
   - view ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.

[ghstack-poisoned]

ailzhang commented on c10/core/TensorImpl.cpp:
```diff
@@ -85,7 +86,7 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
   // We automatically add the corresponding autograd key to key_set_ so that backends can stay
   // in the old way of only registering with backend key like DispatchKey::CPU.
-  if (c10::InferenceMode::is_enabled()) {
+  if (is_view || c10::InferenceMode::is_enabled()) {
```
This logic can be moved to as_view, will come back to this later.

ailzhang replied:
After thinking for a while I decided to keep this logic in constructor for now. One main reason is to keep dispatch_key manipulation of tensor all in one place and local to tensor constructor instead of all over the codebase. But I'm definitely open to suggestions about it.

Commit:
Update on "[WIP]Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view ops.
   - view ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False, view of normal
tensor inside InferenceMode has requires_grad=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. There's no way to change an existing tensor from normal to inference,
or vice versa.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ailzhang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -155,26 +138,28 @@ TEST(InferenceModeTest, TestInferenceTensorInNormalModeInplaceOp) {
 TEST(InferenceModeTest, TestInferenceTensorInNormalModeViewOp) {
   torch::Tensor inference_tensor;
-  for (bool requires_grad: {true, false}) {
+  for (bool requires_grad: {true/*, false*/}) {
```
remove this comment

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False, view of normal
tensor inside InferenceMode has requires_grad=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. There's no way to change an existing tensor from normal to inference,
or vice versa.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False, view of normal
tensor inside InferenceMode has requires_grad=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. There's no way to change an existing tensor from normal to inference,
or vice versa.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor cannot be inplace updated outside InferenceMode.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ailzhang reviewed (commented):
@ezyang @albanD @bhosmer This PR now implements a few relaxed checks we discussed and ready for review when you have time :D . (PR description is also updated with latest behavior).

ezyang commented:
I think you can eliminate this codegen too https://github.com/pytorch/pytorch/pull/53343/files#diff-bb339de87799048156e01dbf0ffb70efe5408f050078521ad02a171fb511c903R345 The inplace case is safe because you added another test on increment version to forbid inplace updates on inference tensors. The view case is safe because we can just generate another inference tensor view, no foul play.

ezyang commented:
I also think it needs to be illegal to `inference_tensor.requires_grad = True` outside of inference mode. (Ideally it would also be illegal inside of inference mode but we have agreed that for implementation reasons we can't do this yet, since there is no way to actually directly allocate a tensor to have `requires_grad=True`. Although I suppose we could have an unsafe "trust me" setter, and then a regular one that tests.)

ezyang commented:
It seems like the "Remove version counter for inference tensors" optimization still hasn't been applied here. Once you do that, that will also naturally cause "Getting version counter" to raise an error, which I think would resolve all of my concerns.

ezyang commented:
PSA for reviewers, I did some major copyediting Ailing's RFC at https://github.com/ailzhang/rfcs/blob/rfc0011/RFC-0011-InferenceMode.md which should accurately describe the state of the world after this PR.

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor cannot be inplace updated outside InferenceMode.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1025,7 +1025,7 @@ Tensor alias_with_sizes_and_strides(
   } else {
     auto impl = c10::make_intrusive<TensorImpl>(
-        Storage(self.storage()), self.key_set(), self.dtype());
+        Storage(self.storage()), self.key_set(), self.dtype(), true);
```
`/* kwarg name */ true` to avoid boolean blindness

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,62 @@ struct C10_API VariableVersion {
   // https://cplusplus.github.io/LWG/issue2334.
   VariableVersion(uint32_t version = 0)
       : version_counter_(c10::make_intrusive<VersionCounter>(version)) {}
+  explicit VariableVersion(bool enabled, uint32_t version = 0) {
```
This is not a very safe constructor, because C++ is willing to convert freely between unsigned int and bool. https://godbolt.org/z/az5jGaofx Additionally, `versions` is ignored when `!enabled`, which makes an illegal state representable.

A better way of doing two variant constructors is to use the scoped enum trick. You can see an example of this in c10/core/DispatchKeySet.h

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -95,6 +97,11 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
   }
 
+  // Inference tensor doesn't have version counter.
+  if (this->is_inference_tensor()) {
```
nit: just `is_inference_tensor()`

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -95,6 +97,11 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
 
+  // Inference tensor doesn't have version counter.
+  if (this->is_inference_tensor()) {
+    version_counter_ = VariableVersion(/*enabled=*/false);
```
I don't know if the compiler was able to figure this out, but what you have in the code here is that first the default constructor for the `version_counter_` field is run (which allocates a new intrusive pointer), and then you immediately afterwards unset it with a nullptr. I suppose you can check what actually happens in the code here with godbolt, but it seems unlikely to me that the compiler will know how to elide the initial allocation if `is_inference_tensor()` is true.

What you should do instead is default initialize the VariableVersion field to nullptr (the cheap initialize), and then only later when you determine that this is an inference tensor.

ezyang replied:
In fact, I would go a step further. Today, we unconditionally allocate a version counter in the constructor for TensorImpl, but in view situations, we then replace it with the shared version counter from the view base. This is also a waste! We could just directly pass in the intended version counter to the TensorImpl constructor. I'm not sure how much refactoring it would take to do this (maybe should have been done before this patch), but it would then make the situation here a lot clearer.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -308,6 +315,8 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
+  TORCH_CHECK(!requires_grad || !is_inference_tensor() || c10::InferenceMode::is_enabled(),
+    "Setting requires_grad=True on inference tensor outside InferenceMode is not allowed.");
```
nit: Probably easier to understand prior to demorganization; `!(requires_grad && is_inference_tensor() && !c10::InferenceMode::is_enabled())` because this matches the English text here (I had to mentally reverse the demorgan rule to verify that this condition was correct)

ezyang replied:
It's probably also worth a comment here saying why doing it inside inference mode is OK

ailzhang replied:
@ezyang For this one I actually want to keep it this way since in most cases we'd expect this check to pass. And the current way allows early escape from the check if any of the first two conditions is true, which will allow us to only hit `c10::InferenceMode::is_enabled()`(pretty expensive TLS read!) when we have `requires_grad=false` inference tensor case. 
But in the easier to understand version we'll always read TLS. Wdyt?

ezyang replied:
It short circuits in either case. `!requires_grad` in an OR statement short circuits when `!requires_grad` is TRUE; while `requires_grad` in an AND statement short circuits when `requires_grad` is FALSE. If you don't believe me try Godbolt for a change https://godbolt.org/z/YPfh5beoM

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1092,21 +1134,27 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   void set_version_counter(
     c10::VariableVersion&& version_counter) noexcept {
```
Since `set_version_counter` may now no longer move out of the rvalue reference here, it is probably clearer to replace this with a pass-by-value signature `set_version_counter(c10::VariableVersion version_counter)`. Otherwise, a call `set_version_counter(std::move(version_counter))` might imply that version counter is dead after the call, but if this is an inference tensor, it's NOT dead, and that would be very surprising to a code reader.

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1026,7 +1026,7 @@ Tensor alias_with_sizes_and_strides(
   } else {
     auto impl = c10::make_intrusive<TensorImpl>(
-        Storage(self.storage()), self.key_set(), self.dtype());
+        Storage(self.storage()), self.key_set(), self.dtype(), /*is_view=*/true);
```
I find it hard to believe that this is the only function that creates views. Even within this file, there are five occurrences of `git grep "Storage(self.storage()"` (and I'm not even sure those are the only occurrences)

ezyang replied:
This is a great opportunity for a staging PR before this one. If you have concluded that there is *no other way* to do what you want to do besides forcing constructors of TensorImpl to declare if they are a view or not, a good, easy prep commit would be to just do that (but maybe not do anything with the information?) and devise a way to validate that you'd gotten all of the necessary sites.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -59,16 +59,18 @@ void TensorImpl::_set_fw_grad(const at::Tensor& new_grad, const at::Tensor& self
 
 TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::TypeMeta data_type,
-                       c10::optional<c10::Device> device_opt)
+                       c10::optional<c10::Device> device_opt, bool is_view)
```
Hmm, stop me if I already commented this. But it seems like there might be a good opportunity to simplify the logic here if you force users to directly pass in the VersionCounter they're supposed to use, in much the same way we force people to pass in the Storage. (Actually, I've always wanted to store the VersionCounter directly on Storage, but because we can detach version counts from their underlying storage they can't actually be stored this way)

ailzhang replied:
@ezyang I definitely agree it's a good opportunity to pass in VersionCounter directly. This improvement is a little tricky tho since we might still want to keep "view tensor share VariableVersion" logic which currently lives in `as_view` still in `as_view` (instead of the TensorImpl constructor which happens inside inference kernel). I might prefer to do that as a followup PR so that we can get this one in first to unblock a few more AutoNonVariableTypeMode replacements. Wdyt?

ezyang replied:
I agree it's tricky. This is definitely one of those cases where I'd personally try to stack a PR before this one doing the refactor to see if it was possible, and if it wasn't but it was feasible in combination with this PR, I'd do it together. Well, re unblocking, let's see if I can convince myself that the way it is currently being done is correct.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,59 @@ struct C10_API VariableVersion {
+  // doesn't have version counter enabled.
   bool unique() const {
-    return 1 == version_counter_.use_count();
+    return this->enabled() ? 1 == version_counter_.use_count() : true;
```
nit: you don't need `this->` here

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,59 @@ struct C10_API VariableVersion {
+      "Inplace update to inference tensor outside InferenceMode is not allowed."
+      "You can make a clone to get a normal tensor before doing inplace update.");
+    if (enabled) {
+      ++version_counter_->version_;
```
Nit, honestly, it's clearer if you just directly test `version_counter_` here, as it makes it clear that you are checking a possibly nullptr before dereferencing it, whereas if you indirect this through the enabled() method it is possible for enabled() to be implemented differently and then the precondition for the pointer dereference no longer holds

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1092,21 +1134,27 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
+  // Inference tensor doesn't have version counter,
+  // set_version_counter is no-op for them.
```
I think you can tighten the invariant here. According to the RFC, inference tensors produce inference views, and normal tensors produce normal views. Thus, an inference tensor which is getting its version counter "set" (we only set version counters when setting up view information) should only ever have gotten the version counter from another inference tensor, which is guaranteed to be disabled. So you should be able to assert that if you ARE an inference tensor, it must be the case that the passed version counter is disabled. And if the passed version counter is disabled, it actually doesn't matter if this is an inference tensor or not, it's always valid to assign the version counter (because overwriting a nullptr with a nullptr does the same thing)--so long as the invariant is upheld.

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -910,6 +910,11 @@ Tensor make_tensor(Args&&... args) {
+template <typename T, typename... Args>
+Tensor make_view_tensor(Args&&... args) {
+  return Tensor(c10::make_intrusive<T>(std::forward<Args>(args)..., /*is_view*/true));
+}
```
Nyeh, I hate this. You can still use `make_tensor` to create a view tensor, and this constructor breaks symmetry with `make_unique` which this function is modeled off of. Remember the enum constructor trick from DispatchKeySet? You can use that here too.

ezyang commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -16,6 +16,24 @@ namespace torch { namespace autograd {
+    // more permissive and only error out when an inference tensor is saved for
+    // backward.  Whether a tensor is saved for backward is determined by derivative
+    // formula and thus varies op by op, so by saying "no inference tensor in autograd"
+    // it's easier for users to follow.
```
This comment is a little out of date. Technically, it's not necessary to do the test here, as given an inference tensor, we will fail when we query `current_version`. However, we do a check here anyway because we can give a better error message in this case.

The comment here also doesn't say why we don't tell users the full story in documentation. It's probably worth explicitly stating the reasoning somewhere too.

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -139,6 +139,14 @@ template<typename... Args> inline variable_list flatten_tensor_args(Args&&... ar
+  // in the default_included_set, so that view ops on inference tensor doesn't
+  // have to go through as_view even outside InferenceMode.
+  if (base.unsafeGetTensorImpl()->is_inference_tensor() || tensor.unsafeGetTensorImpl()->is_inference_tensor())
+    return tensor;
```
I'm not sure why you have to test both base and tensor. The invariant is that inference tensors produce inference views, and normal tensors produce normal views, so if base is an inference tensor, it must necessarily be the case that tensor is an inference tensor. I suppose you could add an assert for this case, but I wouldn't test both branches. (This would simplify the tensorlist code below...)

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -151,10 +148,10 @@ TEST(InferenceModeTest, TestInferenceTensorInNormalModeInplaceOp) {
     {
       InferenceMode guard;
-      inference_tensor = torch::ones({1, 2, 3});
+      inference_tensor = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
```
lol how did we miss this one

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -304,33 +304,54 @@ TEST(InferenceModeTest, TestMixInferenceAndNormalTensorFunctionalOp) {
+    if (requires_grad) {
+      // leaf inference tensor with requires_grad=true can still have gradient.
+      out.backward(torch::ones_like(out));
+      assert_tensor_equal(c.grad(), torch::ones_like(c));
```
ngl this is pretty weird (but maybe ok). One thing to note is that this means that we don't have an invariant that inference tensors are guaranteed not to have AutogradMeta (since the gradient is stored there)

ailzhang replied:
Yea I feel inference tensor eventually can participate in autograd properly as long none of its version/view invariants is violated. Wanted to leave this test here since it's naturally what people would expect from inference tensor participating autograd I think. :)

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -407,3 +431,60 @@ TEST(InferenceModeTest, TestCreationMetaPropagation) {
+  ASSERT_THROWS_WITH(t.unsafeGetTensorImpl()->version_counter().current_version(),
+    "Accessing version_counter of inference tensor is not allowed.");
+  ASSERT_THROWS_WITH(t.unsafeGetTensorImpl()->bump_version(),
+    "Inplace update to inference tensor outside InferenceMode is not allowed.");
```
Thinking about this error message, maybe we can give some clearer guidance about WHY inplace updates to inference tensors outside inference mode are not allowed. Maybe not directly in the error message, but we could possibly link to some documentation which explains the reasoning here (and suggests what people should do instead)

ailzhang replied:
Yup added a link to the rfc for now and a todo to replace it once documentation is available. 
Btw here in the test we're just regexing a part of the error message ;) We do suggest cloning as a workaround in the error message :D

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -407,3 +431,60 @@ TEST(InferenceModeTest, TestCreationMetaPropagation) {
+    t.unsafeGetTensorImpl()->bump_version();
+  }
+  ASSERT_THROWS_WITH(t.unsafeGetTensorImpl()->version_counter().current_version(),
+    "Accessing version_counter of inference tensor is not allowed.");
```
Analogously, we could improve this error message too. In this case, it's easier to say what the problem is (rather than just say "not allowed"); just say inference tensors do not track version counters

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -81,11 +83,12 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
+
+  // Inference tensor doesn't have autograd related keys.
+  // View tensor preserves key_set from their base.
+  if (is_view || inference_mode) {
```
The key thing here, which I didn't get from the comments above, is that if we are constructing a view of an inference tensor, we MUST NOT put autograd key set on the key set because that would result in a non inference tensor view.

In general, this code is all a bit hard to understand because the meaning of `key_set` is overloaded; for non-views, it is guaranteed not to have any autograd keys and we are responsible for adding them; for views, however, we are given a fully fledged key set and are expected to preserve it as much as possible. It seems like it would be easier to understand if we had two separate constructors, one for non-views, and one for views. (This would also nicely solve the `bool is_view` problem that I've been complaining about elsewhere on the PR).

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -401,7 +415,9 @@ void TensorImpl::copy_tensor_metadata(
   copy_tensor_metadata_except_version_counter(src_impl, dest_impl, allow_tensor_metadata_change);
-  dest_impl->set_version_counter(version_counter);
+  if (!dest_impl->is_inference_tensor()) {
+    dest_impl->set_version_counter(version_counter);
```
Perhaps this is another one for subsequent refactoring, but this also looks a bit suspicious, because if I'm copying the VC from source to dest, then if source was an inference tensor the VC is empty and it's harmless to set version counter here. This should be a consequence of https://github.com/pytorch/pytorch/pull/54403/files#r607472380

I think one reason this might not hold is because of this indirect call of copy tensor metadata

```
Tensor VariableHooks::variable_data(const Tensor& self) const {
  TORCH_CHECK(self.defined(), "cannot call variable_data() on undefined tensor");
  auto self_impl_copy = self.unsafeGetTensorImpl()->shallow_copy_and_detach(
    /*version_counter=*/0,
    /*allow_tensor_metadata_change=*/false);
```

so this is freshly creating a new version counter and would trigger the error here. This is kind of suspicious anyway because we don't actually want to allocate a version counter and then throw it out, so maybe just testing if self is an inference tensor in this function would solve this problem.

ailzhang replied:
Yea this one will be considered in the subsequent refactor, essentially we have some code calling `shallow_copy_and_detach` and they're all suspicious. E.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.h#L688 is called when you jit load a model inside InferenceMode and cause here to fail. 
In `as_view` we also call `shallow_copy_and_detach` so the followup refactor will focus on rationalize uses `shallow_copy_and_detach` and the VariableVersions passed from there.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -203,6 +211,12 @@ inline Tensor as_view(const Tensor & base, const Tensor & tensor, bool is_bw_dif
                                    bool is_fw_differentiable, CreationMeta creation_meta=CreationMeta::DEFAULT) {
+  // See Note [View of inference tensor]
+  if (base.unsafeGetTensorImpl()->is_inference_tensor() ||
+        std::all_of(
```
Why is this all and not any to match the or above?

ailzhang replied:
This `std::all` will be removed as we've enforce base and tensor are either both inference or normal. 
I had this check as all mainly because all tensors should have been created at the same time (thus of the same type).

ailzhang commented on c10/core/TensorImpl.cpp:
```diff
@@ -95,6 +98,11 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
 
+  // Inference tensor doesn't have version counter.
+  if (!is_inference_tensor()) {
+    version_counter_ = VariableVersion();
```
version_counter_(VariableVersion(VariableVersion::DISABLED)) does nothing ;) it's here since our default constructor does create intrusive ptr so we want to avoid.

albanD replied:
Well it does create the wrapper object (even though it doesn't create the struct containing the atomic).

ailzhang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -407,3 +431,60 @@ TEST(InferenceModeTest, TestCreationMetaPropagation) {
+    s.add_(t);
+    t.add_(s);
+  }
+  s.copy_(t);
```
Note copy_ is in VariableTypeManual and can be easily ignored if we make some global changes to inplace ops in codegen. In fact I forgot it when I move `increment_version` out of VarialeType kernel :P. So I'm mixing these here to kinda remind we have a manual inplace op for future similar changes.

albanD replied:
Then you should test both?
Right now it is testing half of the feature with add_ and the other half of the feature with copy_ which looks surprising.
Maybe we should test everything with both? :D

ailzhang replied:
sure!

ailzhang commented on c10/core/TensorImpl.cpp:
```diff
@@ -307,7 +315,13 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
+// there's no way that we can directly allocate a tensor to have
+// requires_grad = true in C++ constructor.
 void TensorImpl::set_requires_grad(bool requires_grad) {
+  TORCH_CHECK(!(requires_grad && is_inference_tensor() && !c10::InferenceMode::is_enabled()),
```
In the RFC we decided that we'll forbid explicitly setting requires_grad on inference tensor outside InferenceMode. So this check is to enforce that.

ailzhang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -304,33 +304,54 @@ TEST(InferenceModeTest, TestMixInferenceAndNormalTensorFunctionalOp) {
 
-    ASSERT_THROWS_WITH(torch::add_out(c, s, s), // go through kernels: VariableType(ERROR!), InplaceOrView, CPU
-      "Inference tensor cannot participate in autograd")
+      ASSERT_THROWS_WITH(torch::mul_out(c, s, s), // go through kernels: VariableType(ERROR!), InplaceOrView, CPU
```
Nope, here's testing using inference tensor as out. 
`a` and `s` are both normal tensors so it doesn't make sense to test those?

albanD replied:
Ho in the cpp API, the out arg is the first one? Not the last?

ezyang replied:
yes, this is why they are named `_out`. Faithful API puts them last.

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -10,7 +10,7 @@ inline Tensor view_tensor(
   Storage storage = tensor.storage();
   auto new_tensor = detail::make_tensor<TensorImpl>(
-      std::move(storage), tensor.key_set(), scalarTypeToTypeMeta(dtype));
+      std::move(storage), tensor.key_set(), scalarTypeToTypeMeta(dtype), c10::TensorImpl::VIEW);
```
put it at the front LOL

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -163,14 +160,17 @@ TEST(InferenceModeTest, TestInferenceTensorInNormalModeViewOp) {
+    torch::Tensor out = view_op(inference_tensor); // go through kernels: InplaceOrView, CPU
+    ASSERT_TRUE(is_inference_tensor(out));
+    ASSERT_FALSE(out.requires_grad());
+    ASSERT_FALSE(out.is_view());
```
I guess it's ok to skip all the view logic for these.
This change is starting to get very intrusive and hard to think about...


>  (another reason is if we had ViewMeta in autograd_meta, it suddenly make the tensor requires_grad which is super confusing to users.).

Not sure about that. This is only true if the base does require gradients.

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,58 @@ struct C10_API VariableVersion {
+  // accessed.
+  uint32_t current_version() const {
+    TORCH_CHECK(version_counter_,
+      "Accessing version_counter of inference tensor is not allowed.");
```
A user can do `t._version` in python and will reach this as well ;)

nit: don't specify "version_counter" but just "the version" that way it will work for both the python and c++ API.

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,58 @@ struct C10_API VariableVersion {
 
  public:
+  enum Disabled { DISABLED };
+  // It's okay to return true even for inference tensor which
```
I think it is important to add this to this comment! :)

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -228,20 +229,62 @@ struct C10_API VariableVersion {
   c10::intrusive_ptr<VersionCounter> version_counter_;
 
  public:
+  enum Disabled { DISABLED };
```
I think we are missing a whole paragraph of doc at the begining of version counter to explain what a DISABLED Version counter is.
Also, inference Tensor would be just one use case for this, so maybe we can make all the comments here independent of inference Tensors and just mention that they are one user of the DISABLED flag.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -81,20 +101,27 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
+    // See Note [Expected TLS state in InferenceMode] for why we exclude Autograd & InplaceOrView keys.
+    // Normally key_set only contains backend keys but we do the substraction
+    // here to make sure.
+    key_set_ = key_set - c10::autograd_dispatch_keyset_with_InplaceOrView;
```
nit: as a point of hygiene, if a situation is supposed to be illegal, it's better to ASSERT it, rather than silently work around it, as it will help people identify if something has gone wrong in the system earlier rather than later

ezyang reviewed (approved):
Upcoming refactors will make this better, but I think this is good enough to go today

Commit:
Update on "Relax some limitations of InferenceMode."


A few important points about InferenceMode behavior:
1. All tensors created in InferenceMode are inference tensors except for view of normal tensors.
   - View ops produce output has the same is_inference_tensor property as their input.
     Namely view of normal tensor inside InferenceMode produce a normal tensor, which is
     exactly the same as creating a view inside NoGradMode. And view of
     inference tensor outside InferenceMode produce inference tensor as output.
   - All inference tensors have requires_grad=False and is_leaf=True.
2. All ops are allowed inside InferenceMode, faster than normal mode.
3. Inference tensor cannot be saved for backward.
4. Inference tensor doesn't have version counter.
5. There's no way to change an existing tensor from normal to inference,
or vice versa.
6. Leaf inference tensor with requires_grad=true can still have
gradients.

Differential Revision: [D27316483](https://our.internmc.facebook.com/intern/diff/D27316483)

[ghstack-poisoned]
